File: frontend/src/api/gameArtifact.ts
import { fromBinary } from "@bufbuild/protobuf";
import type { AsyncDuckDBConnection } from "@duckdb/duckdb-wasm";
import {
    Bool,
    Float64,
    Table,
    Uint32,
    Uint64,
    Utf8,
    vectorFromArray,
} from "apache-arrow";

import {
    GameArtifactSchema,
    type GameArtifact,
} from "../generated/artifacts/game_artifact_pb";

export const GAME_ARTIFACT_SCHEMA_VERSION = 1;

const DOWNLOAD_FAILED = "The game artifact could not be downloaded.";
const MISSING_METADATA = "The game artifact has no metadata.";

export const GAME_ARTIFACT_TABLE_NAMES = {
    metadata: "game_metadata",
    plateAppearances: "plate_appearances",
    pitches: "pitches",
} as const;

export type GameArtifactArrowTables = {
    metadata: Table;
    plateAppearances: Table;
    pitches: Table;
};

export type ArtifactConnection = Pick<
    AsyncDuckDBConnection,
    "insertArrowTable"
>;

type Optional<T> = T | null | undefined;

const nullable = <T,>(values: readonly Optional<T>[]): (T | null)[] =>
    values.map((value) => value ?? null);

const uint32Vector = (values: readonly Optional<number>[]) =>
    vectorFromArray(nullable(values), new Uint32());

const uint64Vector = (values: readonly Optional<bigint>[]) =>
    vectorFromArray(nullable(values), new Uint64());

const float64Vector = (values: readonly Optional<number>[]) =>
    vectorFromArray(nullable(values), new Float64());

const booleanVector = (values: readonly Optional<boolean>[]) =>
    vectorFromArray(nullable(values), new Bool());

const stringVector = (values: readonly Optional<string>[]) =>
    vectorFromArray(nullable(values), new Utf8());

export const fetchGameArtifact = async (
    signedUrl: string,
    fetcher: typeof fetch = fetch
): Promise<ArrayBuffer> => {
    if (!signedUrl) throw new Error("The artifact URL is missing.");

    let response: Response;
    try {
        response = await fetcher(signedUrl);
    } catch (cause) {
        throw new Error(DOWNLOAD_FAILED, { cause });
    }

    if (!response.ok) {
        throw new Error(
            `The game artifact download failed with HTTP ${response.status}.`
        );
    }

    try {
        return await response.arrayBuffer();
    } catch (cause) {
        throw new Error(DOWNLOAD_FAILED, { cause });
    }
};

export const decodeGameArtifact = (buffer: ArrayBuffer): GameArtifact => {
    let artifact: GameArtifact;
    try {
        artifact = fromBinary(GameArtifactSchema, new Uint8Array(buffer));
    } catch (cause) {
        throw new Error("The game artifact is not valid Protobuf.", { cause });
    }

    if (artifact.schemaVersion !== GAME_ARTIFACT_SCHEMA_VERSION) {
        throw new Error(
            `Unsupported game artifact schema version ${artifact.schemaVersion}.`
        );
    }
    if (!artifact.metadata) {
        throw new Error(MISSING_METADATA);
    }

    return artifact;
};

export const gameArtifactToArrow = (
    artifact: GameArtifact
): GameArtifactArrowTables => {
    const metadata = artifact.metadata;
    if (!metadata) throw new Error(MISSING_METADATA);

    const appearances = artifact.plateAppearances;
    const pitchRows = appearances.flatMap((appearance) =>
        appearance.pitches.map((pitch) => ({
            gamePk: metadata.gamePk,
            paId: appearance.paId,
            pitch,
        }))
    );

    return {
        metadata: new Table({
            schema_version: uint32Vector([artifact.schemaVersion]),
            game_pk: uint64Vector([metadata.gamePk]),
            official_date: stringVector([metadata.officialDate]),
            game_time_utc: stringVector([metadata.gameTimeUtc]),
            away_team_id: uint64Vector([metadata.awayTeamId]),
            home_team_id: uint64Vector([metadata.homeTeamId]),
            away_team_name: stringVector([metadata.awayTeamName]),
            home_team_name: stringVector([metadata.homeTeamName]),
            away_score: uint32Vector([metadata.awayScore]),
            home_score: uint32Vector([metadata.homeScore]),
            is_doubleheader: booleanVector([metadata.isDoubleheader]),
            game_number: uint32Vector([metadata.gameNumber]),
        }),
        plateAppearances: new Table({
            game_pk: uint64Vector(appearances.map(() => metadata.gamePk)),
            pa_id: uint32Vector(appearances.map((row) => row.paId)),
            batting_team_id: uint64Vector(
                appearances.map((row) => row.battingTeamId)
            ),
            half_inning: uint32Vector(
                appearances.map((row) => row.halfInning)
            ),
            inning: uint32Vector(appearances.map((row) => row.inning)),
            batter_id: uint64Vector(appearances.map((row) => row.batterId)),
            pitcher_id: uint64Vector(appearances.map((row) => row.pitcherId)),
            event: stringVector(appearances.map((row) => row.event)),
            event_type: stringVector(appearances.map((row) => row.eventType)),
            description: stringVector(
                appearances.map((row) => row.description)
            ),
            rbi: uint32Vector(appearances.map((row) => row.rbi)),
            away_score: uint32Vector(
                appearances.map((row) => row.awayScore)
            ),
            home_score: uint32Vector(
                appearances.map((row) => row.homeScore)
            ),
            is_out: booleanVector(appearances.map((row) => row.isOut)),
            balls: uint32Vector(appearances.map((row) => row.balls)),
            strikes: uint32Vector(appearances.map((row) => row.strikes)),
            outs: uint32Vector(appearances.map((row) => row.outs)),
        }),
        pitches: new Table({
            game_pk: uint64Vector(pitchRows.map((row) => row.gamePk)),
            pa_id: uint32Vector(pitchRows.map((row) => row.paId)),
            pitch_number: uint32Vector(
                pitchRows.map((row) => row.pitch.pitchNumber)
            ),
            description: stringVector(
                pitchRows.map((row) => row.pitch.description)
            ),
            code: stringVector(pitchRows.map((row) => row.pitch.code)),
            is_in_play: booleanVector(
                pitchRows.map((row) => row.pitch.isInPlay)
            ),
            is_strike: booleanVector(
                pitchRows.map((row) => row.pitch.isStrike)
            ),
            is_ball: booleanVector(pitchRows.map((row) => row.pitch.isBall)),
            is_out: booleanVector(pitchRows.map((row) => row.pitch.isOut)),
            pitch_type_code: stringVector(
                pitchRows.map((row) => row.pitch.pitchTypeCode)
            ),
            pitch_type_description: stringVector(
                pitchRows.map((row) => row.pitch.pitchTypeDescription)
            ),
            balls: uint32Vector(pitchRows.map((row) => row.pitch.balls)),
            strikes: uint32Vector(pitchRows.map((row) => row.pitch.strikes)),
            outs: uint32Vector(pitchRows.map((row) => row.pitch.outs)),
            pre_balls: uint32Vector(pitchRows.map((row) => row.pitch.preBalls)),
            pre_strikes: uint32Vector(
                pitchRows.map((row) => row.pitch.preStrikes)
            ),
            pre_outs: uint32Vector(pitchRows.map((row) => row.pitch.preOuts)),
            start_speed: float64Vector(
                pitchRows.map((row) => row.pitch.startSpeed)
            ),
            end_speed: float64Vector(
                pitchRows.map((row) => row.pitch.endSpeed)
            ),
            zone: uint32Vector(pitchRows.map((row) => row.pitch.zone)),
            plate_time: float64Vector(
                pitchRows.map((row) => row.pitch.plateTime)
            ),
            extension: float64Vector(
                pitchRows.map((row) => row.pitch.extension)
            ),
            plate_x: float64Vector(pitchRows.map((row) => row.pitch.plateX)),
            plate_z: float64Vector(pitchRows.map((row) => row.pitch.plateZ)),
            spin_rate: uint32Vector(
                pitchRows.map((row) => row.pitch.spinRate)
            ),
            bat_speed: float64Vector(
                pitchRows.map((row) => row.pitch.batSpeed)
            ),
            is_sword_swing: booleanVector(
                pitchRows.map((row) => row.pitch.isSwordSwing)
            ),
            launch_speed: float64Vector(
                pitchRows.map((row) => row.pitch.launchSpeed)
            ),
            launch_angle: float64Vector(
                pitchRows.map((row) => row.pitch.launchAngle)
            ),
            total_distance: float64Vector(
                pitchRows.map((row) => row.pitch.totalDistance)
            ),
            trajectory: stringVector(
                pitchRows.map((row) => row.pitch.trajectory)
            ),
            hardness: stringVector(
                pitchRows.map((row) => row.pitch.hardness)
            ),
            hit_location: stringVector(
                pitchRows.map((row) => row.pitch.hitLocation)
            ),
            hit_coordinate_x: float64Vector(
                pitchRows.map((row) => row.pitch.hitCoordinateX)
            ),
            hit_coordinate_y: float64Vector(
                pitchRows.map((row) => row.pitch.hitCoordinateY)
            ),
            hit_probability: float64Vector(
                pitchRows.map((row) => row.pitch.hitProbability)
            ),
        }),
    };
};

export const registerGameArtifactTables = async (
    connection: ArtifactConnection,
    tables: GameArtifactArrowTables
): Promise<void> => {
    const registrations = [
        [GAME_ARTIFACT_TABLE_NAMES.metadata, tables.metadata],
        [GAME_ARTIFACT_TABLE_NAMES.plateAppearances, tables.plateAppearances],
        [GAME_ARTIFACT_TABLE_NAMES.pitches, tables.pitches],
    ] as const;

    for (const [name, table] of registrations) {
        await connection.insertArrowTable(table, { name, create: true });
    }
};

export const loadGameArtifactIntoDuckDB = async (
    signedUrl: string,
    connection: ArtifactConnection,
    fetcher: typeof fetch = fetch
): Promise<GameArtifact> => {
    const buffer = await fetchGameArtifact(signedUrl, fetcher);
    const artifact = decodeGameArtifact(buffer);
    const tables = gameArtifactToArrow(artifact);
    await registerGameArtifactTables(connection, tables);
    return artifact;
};

File: artifacts/game_artifact.py
"""Build a versioned game artifact from the legacy DuckDB schema."""

from __future__ import annotations

from collections.abc import Iterable
from datetime import UTC, date, datetime
from pathlib import Path
from typing import Any

import duckdb

from artifacts import game_artifact_pb2

SCHEMA_VERSION = 1


def _rows_as_dicts(cursor: duckdb.DuckDBPyConnection) -> Iterable[dict[str, Any]]:
    columns = [column[0] for column in cursor.description]
    return (dict(zip(columns, row, strict=True)) for row in cursor.fetchall())


def _iso_utc(value: datetime | None) -> str:
    if value is None:
        return ""
    if value.tzinfo is None:
        value = value.replace(tzinfo=UTC)
    return value.astimezone(UTC).isoformat().replace("+00:00", "Z")


def _iso_date(value: date | None) -> str:
    return value.isoformat() if value is not None else ""


def _set_present(message: Any, **values: Any) -> None:
    """Set optional protobuf scalars only when DuckDB did not return NULL."""
    for field, value in values.items():
        if value is not None:
            setattr(message, field, value)


def _half_inning(value: str | None) -> int:
    halves = {
        "top": game_artifact_pb2.HALF_INNING_TOP,
        "bottom": game_artifact_pb2.HALF_INNING_BOTTOM,
    }
    return halves.get(value or "", game_artifact_pb2.HALF_INNING_UNSPECIFIED)


def load_game_artifact(
    conn: duckdb.DuckDBPyConnection, game_pk: int
) -> game_artifact_pb2.GameArtifact:
    """Read one legacy game and return its generated protobuf message."""
    game_cursor = conn.execute(
        """
        SELECT
            schedule.gamePk AS game_pk,
            schedule.officialDate AS official_date,
            schedule.gameDate AS game_time_utc,
            info.away_team_id,
            info.home_team_id,
            info.away_team_name,
            info.home_team_name,
            schedule.away_score,
            schedule.home_score,
            schedule.doubleheader AS is_doubleheader,
            schedule.gameNumber AS game_number
        FROM "2025_schedule" AS schedule
        JOIN "2025_game_info" AS info ON info.gamePk = schedule.gamePk
        WHERE schedule.gamePk = ?
        """,
        [game_pk],
    )
    rows = list(_rows_as_dicts(game_cursor))
    if not rows:
        raise LookupError(f"No legacy game with gamePk {game_pk}.")
    if len(rows) != 1:
        raise ValueError(
            f"Expected one legacy row for gamePk {game_pk}, got {len(rows)}."
        )

    row = rows[0]
    artifact = game_artifact_pb2.GameArtifact(schema_version=SCHEMA_VERSION)
    artifact.metadata.game_pk = row["game_pk"]
    artifact.metadata.official_date = _iso_date(row["official_date"])
    artifact.metadata.game_time_utc = _iso_utc(row["game_time_utc"])
    artifact.metadata.away_team_id = row["away_team_id"]
    artifact.metadata.home_team_id = row["home_team_id"]
    artifact.metadata.away_team_name = row["away_team_name"] or ""
    artifact.metadata.home_team_name = row["home_team_name"] or ""
    _set_present(
        artifact.metadata,
        away_score=row["away_score"],
        home_score=row["home_score"],
        is_doubleheader=row["is_doubleheader"],
        game_number=row["game_number"],
    )

    appearances = conn.execute(
        """
        SELECT
            pa_id,
            team_id,
            halfInning AS half_inning,
            inning,
            batter_id,
            pitcher_id,
            event,
            eventType AS event_type,
            description,
            rbi,
            awayScore AS away_score,
            homeScore AS home_score,
            isOut AS is_out,
            balls,
            strikes,
            outs
        FROM "2025_game_plate_appearance"
        WHERE game_pk = ?
        ORDER BY pa_id
        """,
        [game_pk],
    )

    appearances_by_id: dict[int, game_artifact_pb2.PlateAppearance] = {}
    for pa_row in _rows_as_dicts(appearances):
        appearance = artifact.plate_appearances.add(
            pa_id=pa_row["pa_id"],
            batting_team_id=pa_row["team_id"],
            half_inning=_half_inning(pa_row["half_inning"]),
            inning=pa_row["inning"],
            batter_id=pa_row["batter_id"],
            pitcher_id=pa_row["pitcher_id"],
            event=pa_row["event"] or "",
            event_type=pa_row["event_type"] or "",
            description=pa_row["description"] or "",
        )
        _set_present(
            appearance,
            rbi=pa_row["rbi"],
            away_score=pa_row["away_score"],
            home_score=pa_row["home_score"],
            is_out=pa_row["is_out"],
            balls=pa_row["balls"],
            strikes=pa_row["strikes"],
            outs=pa_row["outs"],
        )
        appearances_by_id[pa_row["pa_id"]] = appearance

    pitches = conn.execute(
        """
        SELECT
            pa_id,
            pitch_number,
            description,
            code,
            is_in_play,
            is_strike,
            is_ball,
            is_out,
            pitch_type_code,
            pitch_type_desc AS pitch_type_description,
            balls,
            strikes,
            outs,
            pre_balls,
            pre_strikes,
            pre_outs,
            start_speed,
            end_speed,
            zone,
            plate_time,
            extension,
            pX AS plate_x,
            pZ AS plate_z,
            spin_rate,
            bat_speed,
            is_sword_swing,
            launch_speed,
            launch_angle,
            total_distance,
            trajectory,
            hardness,
            hit_location,
            hit_coord_x AS hit_coordinate_x,
            hit_coord_y AS hit_coordinate_y,
            hit_probability
        FROM pa_plate_events
        WHERE game_pk = ?
        ORDER BY pa_id, pitch_number
        """,
        [game_pk],
    )
    for pitch_row in _rows_as_dicts(pitches):
        pa_id = pitch_row.pop("pa_id")
        try:
            appearance = appearances_by_id[pa_id]
        except KeyError as error:
            raise ValueError(
                f"Pitch {pitch_row['pitch_number']} references missing PA {pa_id}."
            ) from error

        required = {
            "pitch_number": pitch_row.pop("pitch_number"),
            "description": pitch_row.pop("description") or "",
            "code": pitch_row.pop("code") or "",
            "pitch_type_code": pitch_row.pop("pitch_type_code") or "",
            "pitch_type_description": pitch_row.pop("pitch_type_description") or "",
            "trajectory": pitch_row.pop("trajectory") or "",
            "hardness": pitch_row.pop("hardness") or "",
            "hit_location": pitch_row.pop("hit_location") or "",
        }
        pitch = appearance.pitches.add(**required)
        _set_present(pitch, **pitch_row)

    return artifact


def serialize_game(conn: duckdb.DuckDBPyConnection, game_pk: int) -> bytes:
    """Serialize deterministically so checksums are reproducible."""
    return load_game_artifact(conn, game_pk).SerializeToString(deterministic=True)


def deserialize_game(payload: bytes) -> game_artifact_pb2.GameArtifact:
    """Decode through protoc-generated code; no hand-written wire parser is used."""
    return game_artifact_pb2.GameArtifact.FromString(payload)


def write_game_artifact(database: Path, game_pk: int, output: Path) -> int:
    """Export one game and return its serialized byte count."""
    with duckdb.connect(str(database), read_only=True) as conn:
        payload = serialize_game(conn, game_pk)
    output.write_bytes(payload)
    return len(payload)
